File: ForceConnector/ObjectList.vb
Imports System.Data
Imports System.Windows.Forms

Public Class frmObjectList
    Public objectList As List(Of String) = New List(Of String)
    Public langs As List(Of String)
    Public baseLanguage As String = ThisAddIn.userLang

    Public success As Boolean = False

    Dim objects As List(Of Dictionary(Of String, String)) = New List(Of Dictionary(Of String, String))
    Dim standards() As String = {"Account", "Campaign", "Case", "Contact", "Contract", "Event",
        "Lead", "Opportunity", "Pricebook2", "Product2", "Profile", "Quote", "Task", "User", "UserRole"}

    Private Sub frmObjectList_Load(sender As Object, e As EventArgs) Handles Me.Load
        setLanguage()

        If ThisAddIn.usingRESTful Then
            Dim lgr As RESTful.DescribeGlobalSObjectResult() = RESTAPI.getSObjectList()
            With Me.lstObject
                .View = View.Details
                .MultiSelect = True
                .Columns.Add("sObject Name", 200, HorizontalAlignment.Left)
                .Columns.Add("sObject API Name", 265, HorizontalAlignment.Left)
                For Each gr As RESTful.DescribeGlobalSObjectResult In lgr
                    Dim obj = New Dictionary(Of String, String) From {
                    {"name", gr.name},
                    {"label", gr.label},
                    {"custom", gr.custom}
                }
                    objects.Add(obj)
                    If standards.Contains(gr.name) Or gr.custom Then
                        Dim line As New ListViewItem()
                        line.Text = gr.label
                        line.SubItems.Add(gr.name)
                        .Items.Add(line)
                    End If
                Next
                .Sort()
            End With
        Else
            Dim lgr As Partner.DescribeGlobalSObjectResult() = SOAPAPI.getSObjectList()
            With Me.lstObject
                .View = View.Details
                .MultiSelect = True
                .Columns.Add("sObject Name", 200, HorizontalAlignment.Left)
                .Columns.Add("sObject API Name", 265, HorizontalAlignment.Left)
                For Each gr As Partner.DescribeGlobalSObjectResult In lgr
                    Dim obj = New Dictionary(Of String, String) From {
                    {"name", gr.name},
                    {"label", gr.label},
                    {"custom", gr.custom}
                }
                    objects.Add(obj)
                    If standards.Contains(gr.name) Or gr.custom Then
                        Dim line As New ListViewItem()
                        line.Text = gr.label
                        line.SubItems.Add(gr.name)
                        .Items.Add(line)
                    End If
                Next
                .Sort()
            End With
        End If

    End Sub

    Private Sub btnCancel_Click(sender As Object, e As EventArgs) Handles btnCancel.Click
        Me.Close()
    End Sub

    Private Sub btnNext_Click(sender As Object, e As EventArgs) Handles btnNext.Click
        If lstObject.SelectedItems Is Nothing Then
            MessageBox.Show("You must select a object!")
        Else
            For Each itm As ListViewItem In lstObject.SelectedItems
                objectList.Add(itm.SubItems(1).Text)
            Next
            success = True
            Me.Close()
        End If
    End Sub

    Private Sub chkCustom_CheckedChanged(sender As Object, e As EventArgs) Handles chkCustom.CheckedChanged
        resetList()
    End Sub

    Private Sub chkStandard_CheckedChanged(sender As Object, e As EventArgs) Handles chkStandard.CheckedChanged
        resetList()
    End Sub

    Private Sub chkSystem_CheckedChanged(sender As Object, e As EventArgs) Handles chkSystem.CheckedChanged
        resetList()
    End Sub

    Private Sub resetList()
        Dim objs() As Dictionary(Of String, String) = objects.ToArray()
        lstObject.Clear()

        With Me.lstObject
            .View = View.Details
            .MultiSelect = True
            .Columns.Add("sObject Name", 200, HorizontalAlignment.Left)
            .Columns.Add("sObject API Name", 265, HorizontalAlignment.Left)
            For Each obj As Dictionary(Of String, String) In objs
                If chkStandard.Checked And standards.Contains(obj.Item("name")) Then
                    Dim line As New ListViewItem()
                    line.Text = obj.Item("label")
                    line.SubItems.Add(obj.Item("name"))
                    .Items.Add(line)
                End If
                If chkCustom.Checked And obj.Item("custom") = True Then
                    Dim line As New ListViewItem()
                    line.Text = obj.Item("label")
                    line.SubItems.Add(obj.Item("name"))
                    .Items.Add(line)
                End If
                If chkSystem.Checked Then
                    If Not standards.Contains(obj.Item("name")) And obj.Item("custom") = False Then
                        Dim line As New ListViewItem()
                        line.Text = obj.Item("label")
                        line.SubItems.Add(obj.Item("name"))
                        .Items.Add(line)
                    End If
                End If
            Next
            lstObject.Sort()
        End With
    End Sub

    Private Sub setLanguage()
        cmbLang.DisplayMember = "Label"
        cmbLang.ValueMember = "Value"
        Dim langTable As DataTable = New DataTable()
        langTable.Columns.Add("Label", GetType(String))
        langTable.Columns.Add("Value", GetType(String))
        For Each key As String In langs.ToArray()
            langTable.Rows.Add(langSet.Item(key), key)
        Next
        cmbLang.DataSource = langTable
        cmbLang.SelectedItem = ThisAddIn.userLang
    End Sub

    Private Sub cmbLang_SelectedIndexChanged(sender As Object, e As EventArgs) Handles cmbLang.SelectedIndexChanged
        Me.baseLanguage = cmbLang.SelectedValue
    End Sub
End Class

File: ForceConnector/ObjectList.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Windows.Forms;
using Microsoft.VisualBasic.CompilerServices;

namespace ForceConnector
{
    public partial class frmObjectList
    {
        public frmObjectList()
        {
            InitializeComponent();
            _btnNext.Name = "btnNext";
            _btnCancel.Name = "btnCancel";
            _chkStandard.Name = "chkStandard";
            _chkCustom.Name = "chkCustom";
            _chkSystem.Name = "chkSystem";
            _cmbLang.Name = "cmbLang";
        }

        public List<string> objectList = new List<string>();
        public List<string> langs;
        public string baseLanguage = ThisAddIn.userLang;
        public bool success = false;
        private List<Dictionary<string, string>> objects = new List<Dictionary<string, string>>();
        private string[] standards = new string[] { "Account", "Campaign", "Case", "Contact", "Contract", "Event", "Lead", "Opportunity", "Pricebook2", "Product2", "Profile", "Quote", "Task", "User", "UserRole" };

        private void frmObjectList_Load(object sender, EventArgs e)
        {
            setLanguage();
            if (ThisAddIn.usingRESTful)
            {
                var lgr = RESTAPI.getSObjectList();
                {
                    var withBlock = lstObject;
                    withBlock.View = View.Details;
                    withBlock.MultiSelect = true;
                    withBlock.Columns.Add("sObject Name", 200, HorizontalAlignment.Left);
                    withBlock.Columns.Add("sObject API Name", 265, HorizontalAlignment.Left);
                    foreach (RESTful.DescribeGlobalSObjectResult gr in lgr)
                    {
                        var obj = new Dictionary<string, string>() { { "name", gr.name }, { "label", gr.label }, { "custom", Conversions.ToString(gr.custom) } };
                        objects.Add(obj);
                        if (standards.Contains(gr.name) | gr.custom)
                        {
                            var line = new ListViewItem();
                            line.Text = gr.label;
                            line.SubItems.Add(gr.name);
                            withBlock.Items.Add(line);
                        }
                    }

                    withBlock.Sort();
                }
            }
            else
            {
                var lgr = SOAPAPI.getSObjectList();
                {
                    var withBlock1 = lstObject;
                    withBlock1.View = View.Details;
                    withBlock1.MultiSelect = true;
                    withBlock1.Columns.Add("sObject Name", 200, HorizontalAlignment.Left);
                    withBlock1.Columns.Add("sObject API Name", 265, HorizontalAlignment.Left);
                    foreach (Partner.DescribeGlobalSObjectResult gr in lgr)
                    {
                        var obj = new Dictionary<string, string>() { { "name", gr.name }, { "label", gr.label }, { "custom", Conversions.ToString(gr.custom) } };
                        objects.Add(obj);
                        if (standards.Contains(gr.name) | gr.custom)
                        {
                            var line = new ListViewItem();
                            line.Text = gr.label;
                            line.SubItems.Add(gr.name);
                            withBlock1.Items.Add(line);
                        }
                    }

                    withBlock1.Sort();
                }
            }
        }

        private void btnCancel_Click(object sender, EventArgs e)
        {
            Close();
        }

        private void btnNext_Click(object sender, EventArgs e)
        {
            if (lstObject.SelectedItems is null)
            {
                MessageBox.Show("You must select a object!");
            }
            else
            {
                foreach (ListViewItem itm in lstObject.SelectedItems)
                    objectList.Add(itm.SubItems[1].Text);
                success = true;
                Close();
            }
        }

        private void chkCustom_CheckedChanged(object sender, EventArgs e)
        {
            resetList();
        }

        private void chkStandard_CheckedChanged(object sender, EventArgs e)
        {
            resetList();
        }

        private void chkSystem_CheckedChanged(object sender, EventArgs e)
        {
            resetList();
        }

        private void resetList()
        {
            var objs = objects.ToArray();
            lstObject.Clear();
            {
                var withBlock = lstObject;
                withBlock.View = View.Details;
                withBlock.MultiSelect = true;
                withBlock.Columns.Add("sObject Name", 200, HorizontalAlignment.Left);
                withBlock.Columns.Add("sObject API Name", 265, HorizontalAlignment.Left);
                foreach (Dictionary<string, string> obj in objs)
                {
                    if (chkStandard.Checked & standards.Contains(obj["name"]))
                    {
                        var line = new ListViewItem();
                        line.Text = obj["label"];
                        line.SubItems.Add(obj["name"]);
                        withBlock.Items.Add(line);
                    }

                    if (chkCustom.Checked & Conversions.ToBoolean(obj["custom"]) == true)
                    {
                        var line = new ListViewItem();
                        line.Text = obj["label"];
                        line.SubItems.Add(obj["name"]);
                        withBlock.Items.Add(line);
                    }

                    if (chkSystem.Checked)
                    {
                        if (!standards.Contains(obj["name"]) & Conversions.ToBoolean(obj["custom"]) == false)
                        {
                            var line = new ListViewItem();
                            line.Text = obj["label"];
                            line.SubItems.Add(obj["name"]);
                            withBlock.Items.Add(line);
                        }
                    }
                }

                lstObject.Sort();
            }
        }

        private void setLanguage()
        {
            cmbLang.DisplayMember = "Label";
            cmbLang.ValueMember = "Value";
            var langTable = new DataTable();
            langTable.Columns.Add("Label", typeof(string));
            langTable.Columns.Add("Value", typeof(string));
            foreach (string key in langs.ToArray())
                langTable.Rows.Add(DescribeCustomObject.langSet[key], key);
            cmbLang.DataSource = langTable;
            cmbLang.SelectedItem = ThisAddIn.userLang;
        }

        private void cmbLang_SelectedIndexChanged(object sender, EventArgs e)
        {
            baseLanguage = Conversions.ToString(cmbLang.SelectedValue);
        }
    }
}